lib/ecto_job/worker.ex

```diff
@@ -19,14 +19,34 @@ defmodule EctoJob.Worker do
   reactivated by the producer.
   """
   @spec start_link(Config.t, EctoJob.JobQueue.job(), DateTime.t()) :: {:ok, pid}
-  def start_link(config = %Config{repo: repo, execution_timeout: timeout}, job = %queue{}, now) do
-    Task.start_link(fn ->
-      with {:ok, job} <- JobQueue.update_job_in_progress(repo, job, now, timeout) do
-        queue.perform(JobQueue.initial_multi(job), job.params)
-        log_duration(config, job, now)
-        notify_completed(repo, job)
-      end
-    end)
+  def start_link(config, job, now) do
+    Task.start_link(fn -> do_work(config, job, now) end)
+  end
+
+  @spec do_work(Config.t, EctoJob.JobQueue.job(), DateTime.t()) ::
+          :ok | {:ok, EctoJob.JobQueue.job()} | {:error, any()}
+  def do_work(config = %Config{repo: repo,
+                               execution_timeout: exec_timeout,
+                               retrying_timeout: retrying_timeout},
+              job,
+              now) do
+    with {:ok, in_progress_job} <- JobQueue.update_job_in_progress(repo, job, now, exec_timeout),
+         {:ok, _} <- run_queue(config, in_progress_job) do
```

Comment:
I’m not sure we should force jobs to return this, that changes the contract for a job quite a bit, might be a stretch to get into the main repo. Maybe we should just be catching exceptions to trigger the retrying stage?

Reply:
as I understood, even if it's not explicit in the documentation, and maybe I over assumed it, the last call in the job must be `Repo.transaction()` which returns `{:ok, multi}`. I get what you mean though, and I think you are right.  

Maybe we can loosen up this requirement a bit and do the following:

Reply:
```
 with {:ok, in_progress_job} <- JobQueue.update_job_in_progress(repo, job, now, exec_timeout),
         response <- run_queue(config, in_progress_job), 
        true <- valid?(resonse) do
      log_duration(config, in_progress_job, now)
      notify_completed(repo, in_progress_job)
....

def valid?({:error, multi}) do 
   false
end

def valid?(_), do: true
```

Reply:
I think that is pretty important to put it in "RETRYING" if it is a transaction failure, even more important than an exception, wdyt?

Reply:
and sorry for splitting the message in 3, pressed `ctrl enter` by mistake hahahah

Reply:
Ok cool. Let’s see what they say upstream.